sips/sip-018/sip-018-signed-structured-data.md

```diff
@@ -0,0 +1,204 @@
+
+It is important that signed messages are understandable for humans. For transactions, this is obvious: wallet applications display whom is receiving how many tokens of what kind. Likewise, the input parameters, function name, and target contract of contract calls are properly listed. Signed messages that are not transactions should be no different.
+
+The language properties of Clarity make producing and verifying such signed messages intuitive. The specification therefore leverages existing standards and encoding schemes. Messages are represented as native Clarity values and encoded in wire format in preparation for signing. It also makes sure that these signed messages are chain and application specific. There is a focus on ease of verification on the smart contract level.
+
+Below is an example of what a wallet application could show when signing a SIP018 message.
+
+![Wallet SIP018 concept](SIP-018-001.png)
+
+# Specification
+
+The challenge lies in producing messages that are both meaningful to humans as well as easy to be processed on-chain. Luckily, Clarity is a strongly-typed interpreted language. Structured Data is therefore a Clarity Value expression as detailed in *SIP002: Smart Contract Language*. These value expressions are encoded in Stacks wire format, as detailed in *SIP005: Blocks and Transactions*, and then hashed with SHA256. The resulting hash is used as the input for signing.
+
+## Definitions
+
+*Clarity Value*: a native Clarity value as expressed in *SIP002: Smart Contract Language*.
+
+*SHA256*: a cryptographic hash function producing hashes with a length of 32 bytes.
+
+*Signature*: a proof produced by a signing algorithm, in this case ECDSA with the `secp256k1` curve, encoded in RSV order.
+
+*Structured Data*: A structured representation of a message to sign expressed as a Clarity Value.
+
+*Wire format*: the underlying encoding for Clarity Values when generating a transaction, as expressed in *SIP005: Blocks and Transactions*.
+
+## Formal specification
+
+To obtain a signature *S* of structured data *D* using private key *K*:
+
+- `S = signStructuredData(D, K)`; where,
+
+- `signStructuredData(D, K) = sign(messageHash(D), K)`; where `sign` produces a `secp256k1` signature of a message hash using private key *K*; and,
+
+- `messageHash(D) = sha256(structuredDataPrefix || domainHash || structuredDataHash(D))`; where `structuredDataPrefix` is a static value of `"\xC0"`, and `domainHash` and `structuredDataHash` are as described below.
```

Comment:
I'm not convinced that this SIP needs to talk about structured data prefixes or domain hashes as a requirement.  A SIP that's proposing a breaking change (in this case, `to-consensus-buff`) should just focus on the breaking change.  A downstream consumer of `to-consensus-buff` would have its own orthogonal security and namespacing considerations that motivate these two fields.  I'd just drop all discussion of this completely, and put it into a separate SIP.

Reply:
Alternatively, discussion of `to-consensus-buff` could get folded into SIP 015 and dropped here.